timestamp-impl/src/main/java/com/palantir/timestamp/AvailableTimestamps.java

```diff
@@ -0,0 +1,116 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.timestamp;
+
+import static java.util.concurrent.TimeUnit.MINUTES;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.palantir.common.remoting.ServiceNotAvailableException;
+import com.palantir.exception.PalantirInterruptedException;
+
+public class AvailableTimestamps {
+    static final long ALLOCATION_BUFFER_SIZE = 1000 * 1000;
+    private static final long MINIMUM_BUFFER = ALLOCATION_BUFFER_SIZE / 2;
+    private static final long MAX_TIMESTAMPS_TO_HAND_OUT = 10 * 1000;
+
+    private static final Logger log = LoggerFactory.getLogger(AvailableTimestamps.class);
+
+    private final LastReturnedTimestamp lastReturnedTimestamp;
+    private final PersistentUpperLimit upperLimit;
+    private Throwable previousAllocationFailure;
+
+    public AvailableTimestamps(LastReturnedTimestamp lastReturnedTimestamp, PersistentUpperLimit upperLimit) {
+        this.lastReturnedTimestamp = lastReturnedTimestamp;
+        this.upperLimit = upperLimit;
+    }
+
+    public synchronized TimestampRange handOut(long numberToHandOut) {
+        checkArgument(
+                numberToHandOut <= MAX_TIMESTAMPS_TO_HAND_OUT,
+                "Can only hand out %s timestamps at a time, but %s were requested",
+                MAX_TIMESTAMPS_TO_HAND_OUT, numberToHandOut);
+
+        return handOutTimestamp(lastHandedOut() + numberToHandOut);
+    }
+
+    private synchronized TimestampRange handOutTimestamp(long targetTimestamp) {
+        checkArgument(
+                targetTimestamp > lastHandedOut(),
+                "Could not hand out timestamp '%s' as it was earlier than the last handed out timestamp: %s",
+                targetTimestamp, lastHandedOut());
+
+        TimestampRange rangeToHandOut = TimestampRange.createInclusiveRange(lastReturnedTimestamp.get() + 1, targetTimestamp);
+
+        allocateEnoughTimestampsToHandOut(targetTimestamp);
+        lastReturnedTimestamp.increaseToAtLeast(targetTimestamp);
+
+        return rangeToHandOut;
+    }
+
+    private long lastHandedOut() {
+        return lastReturnedTimestamp.get();
+    }
+
+    public synchronized void refreshBuffer() {
+        long buffer = upperLimit.get() - lastHandedOut();
+
+        if (buffer < MINIMUM_BUFFER || !upperLimit.hasIncreasedWithin(1, MINUTES)) {
+            allocateEnoughTimestampsToHandOut(lastHandedOut() + ALLOCATION_BUFFER_SIZE);
+        }
+    }
+
+    public synchronized void fastForwardTo(long newMinimum) {
+        lastReturnedTimestamp.increaseToAtLeast(newMinimum);
+        upperLimit.increaseToAtLeast(newMinimum + ALLOCATION_BUFFER_SIZE);
```

Comment:
Should we be handling an allocation failure here?  I suppose the idea is that this should only ever be an "offline" operation.  If so, can we write that in a comment or javadoc for the method?